### ISSUE TITLE
CA1801 (Review unused parameters) in presence of preprocessor directives

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn-analyzers#1376]
#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Analyzer

ReviewUnusedParametersAnalyzer

#### Repro steps

```
class C
{
    void M(int x)
    {
#if NETSTANDARD
        System.Console.Write(x);
#endif
    }
}
```

#### Expected behavior
No CA1801

#### Actual behavior
```
warning CA1801: Parameter x of method M is never used. Remove the parameter or use it in the method body.
```

**Comment by mavasani** [comment]
This is causing extremely large number of false reports in code written for multi-targeting projects which have lot of preprocessor directives to conditionally compile code for specific frameworks.

**Comment by mavasani** [comment]
This issue also affects remove unused local rule and we may want to update both rules simultaneously to handle these cases.

**Comment by sharwell** [comment]
I would like to get a better understanding of the situation before proceeding with this change.

**Comment by mavasani** [comment]
@jinujoseph @kuhlenh We should probably setup a meeting sometime next week to discuss few of these contentious issues (probably more will come up as I run our analyzers on more code bases).

**Comment by mavasani** [comment]
@sharwell You can see the amount of false positives by turning on this rule on Newtonsoft.json - majority of violations are in this category and something that cannot be fixed.

**Comment by mavasani** [comment]
Our general stance here is that users should also parameterize the presence of the parameter based on the preprocessor directive. So, in the above case:

```csharp
class C
{
    void M(
#if NETSTANDARD
        int x
#endif
     )
    {
#if NETSTANDARD
        System.Console.Write(x);
#endif
    }
}
```
